src/Microsoft.AspNetCore.Authentication.WsFederation/WsFederationHandler.cs

```diff
@@ -164,6 +169,13 @@ public WsFederationHandler(IOptionsMonitor<WsFederationOptions> options, ILogger
                 {
                     return messageReceivedContext.Result;
                 }
+                wsFederationMessage = messageReceivedContext.ProtocolMessage;
+                properties = messageReceivedContext.Properties; // Provides a new instance if not set.
+
+                if (!ValidateCorrelationId(properties) && !Options.AllowUnsolicitedLogins)
```

Comment:
I'd be tempted to think that when you have a non-null `properties` instance (for RP-initiated logins), the correlation identifier should be checked even if `Options.AllowUnsolicitedLogins` was set to `true`.